meta-client/src/main/java/org/apache/helix/metaclient/recipes/leaderelection/LeaderElectionClient.java

```diff
@@ -136,27 +139,63 @@ public void joinLeaderElectionParticipantPool(String leaderPath) {
   public void joinLeaderElectionParticipantPool(String leaderPath, LeaderInfo userInfo) {
     // TODO: create participant entry with info
     subscribeAndTryCreateLeaderEntry(leaderPath);
+
+    LeaderInfo participantInfo = new LeaderInfo(userInfo);
+    createParticipantInfo(leaderPath, participantInfo);
+  }
+
+  private void createParticipantInfo(String leaderPath, LeaderInfo participantInfo) {
+
+    try {
+      // try to create participant info entry, assuming leader election group node is already there
+      _metaClient.create(leaderPath + PARTICIPANTS_ENTRY_PARENT + _participant, participantInfo,
+          MetaClientInterface.EntryMode.EPHEMERAL);
+    } catch (MetaClientNodeExistsException ex) {
+      LOG.info("Already joined pool {}.", leaderPath);
+      return;
+    } catch (MetaClientNoNodeException ex) {
+      try {
+        try {
+          // try to create participants root entry. This function is called after
+          // `subscribeAndTryCreateLeaderEntry`, so leader election root node should have already been created.
+          LOG.info("{} Creating leader group directory {}.", _participant, leaderPath);
+          _metaClient.create(leaderPath + PARTICIPANTS_ENTRY_KEY, null);
+        } catch (MetaClientNodeExistsException ignored) {
+          // participants' root entry created by other client, ignore
+        }
+        // try to participant info node again.
+        _metaClient.create(leaderPath + PARTICIPANTS_ENTRY_PARENT + _participant, participantInfo,
+            MetaClientInterface.EntryMode.EPHEMERAL);
+      } catch (MetaClientNoNodeException e) {
+        // catch exception in previous 2 create()
+        // Leader group root entry or participant parent entry is gone after we checked or created.
+        // Meaning other client removed the group. Throw ConcurrentModificationException.
```

Comment:
Why we need to remove the root dir? Do you mean it destroy the entire leaderelection group? My believe is the entry structure should exist as long as entire leader election group exists. Unless we are destroying everything. Then you should not see the root entry has been deleted case.

Reply:
We will not delete root entry when leader is gone. Actually, we don't event plan to delete root entry even if there is no participant in the leader election pool. 
We still need to protect if the root entry is gone since anyone could delete it using a client.